server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -230,10 +232,15 @@ public SinkQuerySegmentWalker(
                       // 1) Only use caching if data is immutable
                       // 2) Hydrants are not the same between replicas, make sure cache is local
                       if (hydrantDefinitelySwapped && cache.isLocal()) {
+                        StorageAdapter storageAdapter = segmentAndCloseable.lhs.asStorageAdapter();
+                        long segmentMinTime = storageAdapter.getMinTime().getMillis();
+                        long segmentMaxTime = storageAdapter.getMaxTime().getMillis();
+                        Interval actualDataInterval = Intervals.utc(segmentMinTime, segmentMaxTime);
```

Comment:
I think there's an off by one here: `storageAdapter.getMaxTime()` is the highest timestamp in the segment, but the end of `actualDataInterval` is exclusive, and so it doesn't include that highest timestamp. I think doing `segmentMaxTime.plus(1)` will fix it. There's a similar issue in ServerManager.

Reply:
yes, you are right, will fix it